File: csharp/Ba64.cs
using System;
using System.IO;
using System.IO.Compression;

namespace Ba64Lib;

/// <summary>
/// ba64 — binary-to-text encoding that is never larger than standard base64.
/// Reference C# implementation of the v1 format (see ../spec.md). Standard
/// library only (System.IO.Compression, System.Convert). The normative
/// algorithm is spec.md §4. Encoding is a relation, not a function —
/// equality/dedup/MAC comparisons MUST operate on the decoded bytes.
/// </summary>
public static class Ba64
{
    public const long DefaultMaxDecodedLen = 64L << 20; // 64 MiB (spec §7)
    private const byte Version = 0x01;
    private const byte MethodDeflateRaw = 0x01;

    /// <summary>A decode failure carrying a taxonomy code (spec §5).</summary>
    public sealed class Ba64Exception : Exception
    {
        public string Code { get; }
        public Ba64Exception(string code) : base(code) => Code = code;
    }

    // --- CRC-32/ISO-HDLC, hand-rolled table (poly 0xEDB88320) ---------------
    private static readonly uint[] CrcTable = BuildCrcTable();
    private static uint[] BuildCrcTable()
    {
        var t = new uint[256];
        for (uint i = 0; i < 256; i++)
        {
            uint c = i;
            for (int k = 0; k < 8; k++)
                c = (c & 1) != 0 ? 0xEDB88320 ^ (c >> 1) : c >> 1;
            t[i] = c;
        }
        return t;
    }
    private static uint Crc32(ReadOnlySpan<byte> data)
    {
        uint crc = 0xFFFFFFFF;
        foreach (byte b in data)
            crc = CrcTable[(crc ^ b) & 0xff] ^ (crc >> 8);
        return crc ^ 0xFFFFFFFF;
    }

    // --- base64 (canonical, via decode-then-re-encode-and-compare) ----------
    private static byte[] B64CanonicalDecode(string s)
    {
        byte[] raw;
        try { raw = Convert.FromBase64String(s); }
        catch (FormatException) { throw new Ba64Exception("E_BASE64"); }
        if (Convert.ToBase64String(raw) != s) throw new Ba64Exception("E_BASE64");
        return raw;
    }

    // --- LEB128 -------------------------------------------------------------
    private static byte[] Leb128Encode(ulong n)
    {
        Span<byte> tmp = stackalloc byte[10];
        int i = 0;
        while (true)
        {
            byte b = (byte)(n & 0x7f);
            n >>= 7;
            if (n != 0) tmp[i++] = (byte)(b | 0x80);
            else { tmp[i++] = b; return tmp[..i].ToArray(); }
        }
    }
    private static (ulong value, int pos) Leb128Decode(byte[] buf, int pos)
    {
        ulong value = 0; int shift = 0, start = pos;
        while (true)
        {
            if (pos >= buf.Length) throw new Ba64Exception("E_TRUNCATED");
            int b = buf[pos];
            pos++;
            if (pos - start > 9) throw new Ba64Exception("E_HEADER");
            value |= (ulong)(b & 0x7f) << shift;
            if ((b & 0x80) == 0)
            {
                if (pos - start > 1 && b == 0) throw new Ba64Exception("E_HEADER"); // non-minimal
                return (value, pos);
            }
            shift += 7;
        }
    }

    // Lazily yields one byte per Read so DeflateStream consumes exactly the
    // stream (no over-read), letting us detect trailing bytes precisely.
    private sealed class LazyStream : Stream
    {
        private readonly byte[] _d;
        public int Consumed;
        public LazyStream(byte[] d) => _d = d;
        public override int Read(byte[] b, int o, int c)
        {
            if (Consumed >= _d.Length || c == 0) return 0;
            b[o] = _d[Consumed++];
            return 1;
        }
        public override bool CanRead => true;
        public override bool CanSeek => false;
        public override bool CanWrite => false;
        public override long Length => _d.Length;
        public override long Position { get => Consumed; set { } }
        public override void Flush() { }
        public override long Seek(long a, SeekOrigin b) => 0;
        public override void SetLength(long a) { }
        public override void Write(byte[] a, int b, int c) { }
    }

    private static byte[] InflateExact(byte[] payload, long decodedLen)
    {
        if (payload.Length == 0) throw new Ba64Exception("E_PAYLOAD"); // no DEFLATE stream at all
        long bound = (long)payload.Length * 1032 + 64;
        // A real DEFLATE stream can't produce more than `bound`, so the buffer
        // never needs to exceed it even when decodedLen is larger.
        int cap = (int)Math.Min(Math.Min(decodedLen, bound) + 1, int.MaxValue);
        var baseStream = new LazyStream(payload);
        var outBuf = new byte[cap];
        int total = 0;
        try
        {
            using var inf = new DeflateStream(baseStream, CompressionMode.Decompress);
            while (total < outBuf.Length)
            {
                int n = inf.Read(outBuf, total, outBuf.Length - total);
                if (n == 0) break;
                total += n;
            }
        }
        catch (InvalidDataException) { throw new Ba64Exception("E_PAYLOAD"); } // malformed/truncated
        if (total > decodedLen) throw new Ba64Exception("E_LENGTH_MISMATCH"); // exceeded cap
        if (baseStream.Consumed != payload.Length) throw new Ba64Exception("E_PAYLOAD"); // trailing bytes
        return outBuf[..total];
    }

    // --- public API ---------------------------------------------------------
    public static string Encode(byte[] data) => EncodeLevel(data, 6);

    public static string EncodeLevel(byte[] data, int level)
    {
        string plain = Convert.ToBase64String(data);
        var cl = level <= 0 ? CompressionLevel.NoCompression
               : level <= 3 ? CompressionLevel.Fastest
               : level <= 6 ? CompressionLevel.Optimal
               : CompressionLevel.SmallestSize;
        byte[] payload;
        using (var ms = new MemoryStream())
        {
            using (var d = new DeflateStream(ms, cl, true)) d.Write(data, 0, data.Length);
            payload = ms.ToArray();
        }
        using var frame = new MemoryStream();
        frame.WriteByte(Version);
        frame.WriteByte(MethodDeflateRaw);
        var len = Leb128Encode((ulong)data.Length);
        frame.Write(len, 0, len.Length);
        uint crc = Crc32(data);
        frame.WriteByte((byte)(crc & 0xff));
        frame.WriteByte((byte)(crc >> 8 & 0xff));
        frame.WriteByte((byte)(crc >> 16 & 0xff));
        frame.WriteByte((byte)(crc >> 24 & 0xff));
        frame.Write(payload, 0, payload.Length);

        string candidate = "=" + Convert.ToBase64String(frame.ToArray());
        return candidate.Length < plain.Length ? candidate : plain;
    }

    public static byte[] Decode(string text) => DecodeMax(text, DefaultMaxDecodedLen);

    public static byte[] DecodeMax(string text, long maxDecodedLen)
    {
        if (maxDecodedLen < 0) maxDecodedLen = 0; // a negative limit must not wrap to a huge ulong
        if (text.Length == 0 || text[0] != '=') return B64CanonicalDecode(text); // step 1

        byte[] frame = B64CanonicalDecode(text[1..]); // step 2

        if (frame.Length < 1) throw new Ba64Exception("E_TRUNCATED"); // step 3
        if (frame[0] != Version) throw new Ba64Exception("E_VERSION");
        if (frame.Length < 2) throw new Ba64Exception("E_TRUNCATED"); // step 4
        if (frame[1] != MethodDeflateRaw) throw new Ba64Exception("E_METHOD");

        var (value, pos) = Leb128Decode(frame, 2); // step 5
        if (value > (ulong)maxDecodedLen) throw new Ba64Exception("E_LIMIT_EXCEEDED"); // step 6
        long decodedLen = (long)value; // keep full width; a 32-bit cast would silently mis-length

        if (frame.Length - pos < 4) throw new Ba64Exception("E_TRUNCATED"); // step 7
        uint crcStored = (uint)(frame[pos] | frame[pos + 1] << 8 | frame[pos + 2] << 16 | frame[pos + 3] << 24);
        byte[] payload = frame[(pos + 4)..];

        byte[] outBytes = InflateExact(payload, decodedLen); // step 8
        if (outBytes.Length != decodedLen) throw new Ba64Exception("E_LENGTH_MISMATCH");

        if (Crc32(outBytes) != crcStored) throw new Ba64Exception("E_CHECKSUM"); // step 9
        return outBytes; // step 10
    }
}

File: java/Ba64.java
import java.util.Base64;
import java.util.zip.CRC32;
import java.util.zip.DataFormatException;
import java.util.zip.Deflater;
import java.util.zip.Inflater;

/**
 * ba64 — binary-to-text encoding that is never larger than standard base64.
 *
 * Reference Java implementation of the v1 format (see ../spec.md). Standard
 * library only (java.util.zip, java.util.Base64). The normative algorithm is
 * spec.md §4. Encoding is a relation, not a function — equality/dedup/MAC
 * comparisons MUST operate on the decoded bytes.
 */
public final class Ba64 {
    public static final long DEFAULT_MAX_DECODED_LEN = 64L << 20; // 64 MiB (spec §7)

    private static final int VERSION = 0x01;
    private static final int METHOD_DEFLATE_RAW = 0x01;

    private Ba64() {}

    /** A decode failure carrying a machine-readable taxonomy code (spec §5). */
    public static final class Ba64Exception extends RuntimeException {
        private final String code;
        public Ba64Exception(String code) { super(code); this.code = code; }
        public String code() { return code; }
    }

    private static final Base64.Encoder B64E = Base64.getEncoder();
    private static final Base64.Decoder B64D = Base64.getDecoder();

    /** Canonical RFC 4648 §4 base64 decode, else E_BASE64. Java's decoder is
     * lenient about padding/trailing bits, so strictness is by
     * decode-then-re-encode-and-compare (spec §4 note). */
    private static byte[] b64CanonicalDecode(String s) {
        byte[] raw;
        try {
            raw = B64D.decode(s);
        } catch (IllegalArgumentException e) {
            throw new Ba64Exception("E_BASE64");
        }
        if (!B64E.encodeToString(raw).equals(s)) throw new Ba64Exception("E_BASE64");
        return raw;
    }

    private static byte[] leb128Encode(long n) {
        byte[] tmp = new byte[10];
        int i = 0;
        while (true) {
            int b = (int) (n & 0x7f);
            n >>>= 7;
            if (n != 0) tmp[i++] = (byte) (b | 0x80);
            else { tmp[i++] = (byte) b; byte[] out = new byte[i]; System.arraycopy(tmp, 0, out, 0, i); return out; }
        }
    }

    /** Returns {value, newPos} packed as long[]{value, pos}. */
    private static long[] leb128Decode(byte[] buf, int pos) {
        long value = 0;
        int shift = 0, start = pos;
        while (true) {
            if (pos >= buf.length) throw new Ba64Exception("E_TRUNCATED");
            int b = buf[pos] & 0xff;
            pos++;
            if (pos - start > 9) throw new Ba64Exception("E_HEADER");
            value |= (long) (b & 0x7f) << shift;
            if ((b & 0x80) == 0) {
                if (pos - start > 1 && b == 0) throw new Ba64Exception("E_HEADER"); // non-minimal
                return new long[] { value, pos };
            }
            shift += 7;
        }
    }

    /** Inflate raw DEFLATE with a hard output cap; detect exact end and trailing
     * bytes. Cap allocation at what DEFLATE could produce from the payload
     * (~1032x+64) so a small frame claiming a huge size stays cheap (spec §7). */
    private static byte[] inflateExact(byte[] payload, long decodedLen) {
        long bound = (long) payload.length * 1032 + 64;
        // A real DEFLATE stream can't produce more than `bound` bytes, so the
        // output buffer never needs to exceed it even when decodedLen is larger.
        int cap = (int) Math.min(Math.min(decodedLen, bound) + 1, Integer.MAX_VALUE);
        byte[] out = new byte[cap];
        Inflater inf = new Inflater(true); // nowrap = raw DEFLATE
        inf.setInput(payload);
        int total = 0;
        try {
            while (!inf.finished()) {
                if (total == out.length) throw new Ba64Exception("E_LENGTH_MISMATCH"); // exceeded cap
                int n = inf.inflate(out, total, out.length - total);
                if (n > 0) {
                    total += n;
                    if (total > decodedLen) throw new Ba64Exception("E_LENGTH_MISMATCH");
                } else if (inf.finished()) {
                    break;
                } else {
                    throw new Ba64Exception("E_PAYLOAD"); // needsInput (truncated) or no progress
                }
            }
        } catch (DataFormatException e) {
            throw new Ba64Exception("E_PAYLOAD"); // malformed stream
        }
        if (inf.getRemaining() != 0) throw new Ba64Exception("E_PAYLOAD"); // trailing bytes
        inf.end();
        byte[] result = new byte[total];
        System.arraycopy(out, 0, result, 0, total);
        return result;
    }

    public static String encode(byte[] data) { return encode(data, 6); }

    public static String encode(byte[] data, int level) {
        String plain = B64E.encodeToString(data);

        Deflater def = new Deflater(level, true); // nowrap = raw DEFLATE
        def.setInput(data);
        def.finish();
        byte[] buf = new byte[Math.max(64, data.length + 64)];
        java.io.ByteArrayOutputStream payload = new java.io.ByteArrayOutputStream();
        while (!def.finished()) {
            int n = def.deflate(buf);
            payload.write(buf, 0, n);
        }
        def.end();

        CRC32 crc = new CRC32();
        crc.update(data);
        long crcVal = crc.getValue();

        java.io.ByteArrayOutputStream frame = new java.io.ByteArrayOutputStream();
        frame.write(VERSION);
        frame.write(METHOD_DEFLATE_RAW);
        byte[] len = leb128Encode(data.length);
        frame.write(len, 0, len.length);
        frame.write((int) (crcVal & 0xff));
        frame.write((int) (crcVal >>> 8 & 0xff));
        frame.write((int) (crcVal >>> 16 & 0xff));
        frame.write((int) (crcVal >>> 24 & 0xff));
        byte[] p = payload.toByteArray();
        frame.write(p, 0, p.length);

        String candidate = "=" + B64E.encodeToString(frame.toByteArray());
        return candidate.length() < plain.length() ? candidate : plain;
    }

    public static byte[] decode(String text) { return decode(text, DEFAULT_MAX_DECODED_LEN); }

    public static byte[] decode(String text, long maxDecodedLen) {
        if (text.isEmpty() || text.charAt(0) != '=') return b64CanonicalDecode(text); // step 1

        byte[] frame = b64CanonicalDecode(text.substring(1)); // step 2

        if (frame.length < 1) throw new Ba64Exception("E_TRUNCATED"); // step 3
        if ((frame[0] & 0xff) != VERSION) throw new Ba64Exception("E_VERSION");
        if (frame.length < 2) throw new Ba64Exception("E_TRUNCATED"); // step 4
        if ((frame[1] & 0xff) != METHOD_DEFLATE_RAW) throw new Ba64Exception("E_METHOD");

        long[] vp = leb128Decode(frame, 2); // step 5
        long value = vp[0];
        int pos = (int) vp[1];

        if (Long.compareUnsigned(value, maxDecodedLen) > 0) throw new Ba64Exception("E_LIMIT_EXCEEDED"); // step 6
        long decodedLen = value; // keep full width; a 32-bit cast would silently mis-length

        if (frame.length - pos < 4) throw new Ba64Exception("E_TRUNCATED"); // step 7
        long crcStored = (frame[pos] & 0xffL) | (frame[pos + 1] & 0xffL) << 8
                | (frame[pos + 2] & 0xffL) << 16 | (frame[pos + 3] & 0xffL) << 24;
        byte[] payload = new byte[frame.length - pos - 4];
        System.arraycopy(frame, pos + 4, payload, 0, payload.length);

        byte[] out = inflateExact(payload, decodedLen); // step 8
        if (out.length != decodedLen) throw new Ba64Exception("E_LENGTH_MISMATCH");

        CRC32 crc = new CRC32();
        crc.update(out);
        if (crc.getValue() != crcStored) throw new Ba64Exception("E_CHECKSUM"); // step 9
        return out; // step 10
    }
}
